activerecord/lib/active_record/connection_adapters/postgresql/quoting.rb

```diff
@@ -97,7 +97,7 @@ def quote_table_name_for_assignment(table, attr)
 
         # Quotes column names for use in SQL queries.
         def quote_column_name(name) # :nodoc:
-          self.class.quoted_column_names[name] ||= PG::Connection.quote_ident(super).freeze
+          self.class.quoted_column_names.compute_if_absent(name) { PG::Connection.quote_ident(super).freeze }
```

Comment:
So I really don't care if there is a race here, since both threads will compute the same result, so it just means we'll compute the value twice instead of once, no big deal.

So I don't think `compute_if_absent` is justified here. I'd rather just use `Concurrent::Hash` rather than `Map` and keep the old `[]name| |=`.

Reply:
Other than that, I'm OK with the change, as it would have no impact on MRI.

Reply:
What worries me a bit more is that there is a comment in `Concurrent::Hash` that suggest it might get a Mutex on MRI: https://github.com/ruby-concurrency/concurrent-ruby/blob/3bfe39e77958876403cefa3a2d89e86c7bdb5f9d/lib/concurrent-ruby/concurrent/hash.rb#L18-L23

And that would suck.

Reply:
I think Concurrent::Map is the right collection to use here.

One huge problem with Concurrent::Hash is it does not scale even for reads on non-GVL Rubies, which means there can very easily be contention even when the cache if fully populated.

Perf-wise, Concurrent::Map#[] is [very close to](https://github.com/ruby-concurrency/concurrent-ruby/issues/882#issuecomment-1645802440) `Concurrent::Hash#[]` on CRuby, the Mutex is [only used when writing](https://github.com/ruby-concurrency/concurrent-ruby/blob/3bfe39e77958876403cefa3a2d89e86c7bdb5f9d/lib/concurrent-ruby/concurrent/collection/map/mri_map_backend.rb).

In fact as a concurrent-ruby maintainer I want to discourage to use Concurrent::Hash, it makes no sense to preserve ordering for a hash/map accessed concurrently, that kills scalability, and it doesn't have the proper methods for concurrency like `compute_if_absent`. I'm thinking it might be worth deprecating it even for that reason.

`compute_if_absent` is [already well optimized](https://github.com/ruby-concurrency/concurrent-ruby/blob/3bfe39e77958876403cefa3a2d89e86c7bdb5f9d/lib/concurrent-ruby/concurrent/collection/map/mri_map_backend.rb#L22) for Concurrent::Map.
One could still use `||=` with Concurrent::Map if they like, but it would be best to measure if it's really faster then before giving up on extra work (including on CRuby).

Reply:
> One huge problem with Concurrent::Hash is it does not scale even for reads on non-GVL Rubies

So my personal stance is that I'm happy to merge changes in Rails that help compatibility with other Ruby implementations, but never at the detriment of performance on MRI. Hence I'd like this cache to remain a `::Hash` on MRI.

Also I don't want a lock to be held when calling the `quote` function. Because right now we use the static method, but we may need to call `quote_ident` on the connection in the future for it to be encoding aware, so it could lead to contention between different connections, so IMO `compute_if_absent` isn't a good fit here.

Reply:
> I think the difference is very small

I just don't want to have to synchronize a mutex if I know it's useless.

> it will be exactly the same between Concurrent::Hash and Concurrent::Map when I add a write lock for Concurrent::Hash

Then I'll refactor Rails to stop using it on MRI.

Reply:
> I just don't want to have to synchronize a mutex if I know it's useless.

`Concurrent::Map#[]` doesn't use a Mutex, it's literally just this:
https://github.com/ruby-concurrency/concurrent-ruby/blob/3bfe39e77958876403cefa3a2d89e86c7bdb5f9d/lib/concurrent-ruby/concurrent/collection/map/non_concurrent_map_backend.rb#L21-L23

Reply:
`Concurrent::Map#[]=` does https://github.com/ruby-concurrency/concurrent-ruby/blob/3bfe39e77958876403cefa3a2d89e86c7bdb5f9d/lib/concurrent-ruby/concurrent/collection/map/mri_map_backend.rb#L17-L19

Reply:
Yes, but as every cache we expect far more reads than writes, isn't it?
I doubt the Mutex there is a big overhead over calling the block + allocations + the work `PG::Connection.quote_ident(super).freeze` does.

Reply:
Again, I just don't want any extra overhead, however how small, when 99.99% of our users don't need it. That shouldn't even be remotely controversial. Yes it's not a major overhead, but these things add up and end in a death by a thousand cuts.

But no big deal, I'll add an indirection in Active Support, e.g. `ActiveSupport::GVLHash` which is just `::Hash` on MRI and `Concurrent::Map` on other Rubies.